### PR TITLE
[Feature](bangc-ops): add mutual_information_forward op

### PR DESCRIPTION
**Pull request by tcd-fang**
Thanks for your contribution and we appreciate it a lot. 

## 1. Motivation

add new_op, mutual_information_forward

## 2. Modification

```
new file:   bangc-ops/kernels/mutual_information_forward/mutual_information_forward.cpp
new file:   bangc-ops/kernels/mutual_information_forward/mutual_information_forward.h
new file:   bangc-ops/kernels/mutual_information_forward/mutual_information_forward_block.mlu
new file:   bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/mutual_information_forward/mutual_information_forward.cpp
new file:   bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/mutual_information_forward/mutual_information_forward.h
new file:   bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/mutual_information_forward/test_case/mutual_information_forward.prototxt
new file:   docs/bangc-docs/design_docs/mutual_information_forward/mutual_information_forward.md
modified:   bangc-ops/kernels/mutual_information_backward/mutual_information_backward.cpp
modified:   bangc-ops/kernels/mutual_information_backward/mutual_information_backward_block.mlu
modified:   bangc-ops/mlu_op.h
modified:   docs/bangc-docs/user_guide/9_operators/index.rst
```

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- [x] diff1: diff1 <= 3e-3
- [x] diff2: diff2 <= 3e-3

#### 3.1.2 Operator Scheme checklist

|     No.        |                 Details              |            Check Results             |
|----------------|--------------------------------------|--------------------------------------|
|        1       |Supported hardware                    |             MLU370<br>MLU590         |
|        2       |Job types                             |          block     |
|        3       |Layouts                               |          ARRAY      |
|        4       |Whether multi-dimensions are supported|             no              |
|        5       |Whether element zero is supported     |                 yes                |
|        6       |Data type(half/float)                 |           float           |
|        7       |Whether there is size limit           |       The shape of “py” is  [B, S + 1, T]<br>On MLU370:<br>T * (S + 1) + (T + 1) * S + (T + 1) * (S + 1) + 9 * std::min(S, T) + 11 <= 163840.<br>On MLU590:<br/>T * (S + 1) + (T + 1) * S + (T + 1) * (S + 1) + 9 * std::min(S, T) + 11 <= 98304.        |

#### 3.1.3 New Feature Test

If you have checked the following items, please tick the relevant box.

- [x] Data type test
- [ ] Multi-dimensional tensor test
- [ ] Layout test
- [x] Different size/integer remainder end segment/alignment misalignment test
- [x] Zero dimensional tensor test/zero element test
- [x] stability test
- [x] Multiple platform test
- [x] Gen_case module test
- [x] Nan/INF tests 
- [ ] Bug fix tests
- [x] For memory leak check details, see[GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [x] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [x] For I/O calculation efficiency check details, see: [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md).

#### 3.1.4 Parameter Check

When a new operator is submitted, the test points are given and the test results are stated.

|                   Test Point                    | Acceptance Standard | Test Result (Error Message) |
| ----------------------------------------------- | --------------------| --------------------------- |
| Whether it conforms to the operator restriction |     Normal error    |      pass           |
| Whether illegal parameters are passed           |     Normal error    |        pass           |

### 3.2 Accuracy Test

For the cases used in the New Feature Test section, the features and the number of cases are recorded here. When multiple operations are tested, multiple tables are needed to include details of these operations.

Operation:

| Test Point          | Description                        | Quantity | Comment                                                      |
| ------------------- | ---------------------------------- | -------- | ------------------------------------------------------------ |
| Data type test      | float                              | pass     | [       OK ] mutual_information_forward/TestSuite.mluOp/27 (22 ms)<br/>[----------] 28 tests from mutual_information_forward/TestSuite (1084 ms total)<br/><br/>[----------] Global test environment tear-down<br/>[ SUMMARY  ] Total 28 cases of 1 op(s).<br/>ALL PASSED.<br/>[==========] 28 test cases from 1 test suite ran. (1434 ms total)<br/>[  PASSED  ] 28 test cases. |
| Zero element test   | Whether to support this test       | pass     | [2023-5-18 14:37:50] [MLUOP] [Vlog]:MutualInformationForwardExecutor compute.<br/>[2023-5-18 14:37:50] [MLUOP] [Vlog]:[mluOpMutualInformationForward] Skip zero element tensor when px.shape[0] is zero.<br/>[2023-5-18 14:37:50] [MLUOP] [Vlog]:Device free (for workspace). |
| Stability test      | --gtest_repeat=100<br>--thread=100 | pass     | ./mluop_gtest --gtest_filter=*mutual_information_forward* --cases_dir=/projs/cnnl/tangchengda/test_cases  --gtest_repeat=100 --thread=100<br>[----------] Global test environment tear-down<br/>[ SUMMARY  ] Total 2800 cases of 100 op(s).<br/>ALL PASSED.<br/>[==========] 1 test case from 1 test suite ran. (466 ms total)<br/>[  PASSED  ] 1 test case. |
| Mult-platform test  | MLU370/MLU590                      | pass     | MLU370X4:<br>[       OK ] mutual_information_forward/TestSuite.mluOp/27 (22 ms)<br/>[----------] 28 tests from mutual_information_forward/TestSuite (1084 ms total)<br/><br/>[----------] Global test environment tear-down<br/>[ SUMMARY  ] Total 28 cases of 1 op(s).<br/>ALL PASSED.<br/>[==========] 28 test cases from 1 test suite ran. (1434 ms total)<br/>[  PASSED  ] 28 test cases.<br>MLU590:<br/>[       OK ] mutual_information_forward/TestSuite.mluOp/27 (19 ms)<br/>[----------] 28 tests from mutual_information_forward/TestSuite (1049 ms total)<br/><br/>[----------] Global test environment tear-down<br/>[ SUMMARY  ] Total 28 cases of 1 op(s).<br/>ALL PASSED.<br/>[==========] 28 test cases from 1 test suite ran. (1541 ms total)<br/>[  PASSED  ] 28 test cases. |
| Nan/INF test        | Whether to support this test       | pass     | [ SUMMARY  ] Total 50 cases of 1 op(s).<br/>ALL PASSED.<br/>[==========] 50 test cases from 1 test suite ran. (2910 ms total)<br/>[  PASSED  ] 50 test cases. |
| Memory leak check   | Test result                        | pass     | [ SUMMARY  ] Total 28 cases of 1 op(s).<br/>ALL PASSED.<br/>[==========] 28 test cases from 1 test suite ran. (1434 ms total)<br/>[  PASSED  ] 28 test cases. |
| Code coverage check | Test result                        | pass     | File - Line Coverage - Functions<br>mutual_information_forward_block.mlu - [98.8% 242 / 245] - [100.0% 8 / 8] |

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|-----|----|----|----|-----|
|mutual_information_forward| 138 | 64.126 | 0.00249094 | 0.000272899 | -1 | float | px:[4, 15, 105], py:[4, 16, 104], p:[4, 16, 105] |

Platform：MLU590

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|-----|----|----|----|-----|
|mutual_information_forward| 126 | 18.804 | 0.000303131 | 0.000161898 | -1 | float | px:[4, 15, 105], py:[4, 16, 104], p:[4, 16, 105] |

### 3.4 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

File: bangc-ops/mlu_op.h [hunk 1]
```diff
@@ -7698,6 +7699,167 @@ mluOpMutualInformationBackward(mluOpHandle_t handle,
...
+ * @param[in] opt_boundary
+ * Pointer to the MLU memory that stores the \b opt_boundary tensor.
+ * @param[in] p_desc
+ * The descriptor of the tensor \b p, which contains dimension, data type, and data layout.
+ * For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] p
+ * Pointer to the MLU memory that stores the tensor \b p.
+ * @param[in] workspace
+ * Pointer to the MLU memory that is used as an extra workspace for the mutual_information_forward operation.
+ * For more information about workspace, see "Cambricon BANGC OPS User Guide".
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:7791]:
```suggestion
 * For more information about the workspace, see "Cambricon BANG C OPS User Guide".
```

File: bangc-ops/mlu_op.h [hunk 2]
```diff
@@ -7698,6 +7699,167 @@ mluOpMutualInformationBackward(mluOpHandle_t handle,
...
+ *   - tensor ``opt_boundary`` : ``int64``;
+ *   - tensor ``p`` : ``float``;
+ *   - tensor ``ans`` : ``float``;
+ *
+ * @par Data Layout
+ * - The supported layout of input and output tensors must be \p MLUOP_LAYOUT_ARRAY.
+ *
+ * @par Scale Limitation
+ * - The shape of \b px is 3D([B, S, T + 1]), where B is the batch size, S is the length of symbols
+ *   and T is the length of input.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:7816]:
```suggestion
 *   and T is the length of the input.
```

File: bangc-ops/mlu_op.h [hunk 3]
```diff
@@ -7698,6 +7699,167 @@ mluOpMutualInformationBackward(mluOpHandle_t handle,
...
+// Group:MutualInformationForward
+/*!
+ * @brief Returns the size of the MLU memory that is used as an extra workspace
+ * to optimize ::mluOpMutualInformationForward.
+ *
+ * @param[in] handle
+ * Handle to an MLUOP context that is used to manage MLU devices and queues in
+ * ::mluOpMutualInformationForward. For detailed information, see ::mluOpHandle_t.
+ * @param[in] px_desc
+ * The descriptor of the tensor \b px, which contains dimension, data type, and data layout.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:7711]: 通常短语不能接定语从句
```suggestion
 * The descriptor of the tensor \b px containing dimension, data type, and data layout.
```

或者

```suggestion
 * The descriptor of the tensor \b px. It contains dimension, data type, and data layout.
```

File: bangc-ops/mlu_op.h [hunk 4]
```diff
@@ -7698,6 +7699,167 @@ mluOpMutualInformationBackward(mluOpHandle_t handle,
...
+ *
+ *     - T * (S + 1) + (T + 1) * S + (T + 1) * (S + 1) + 9 * min(S, T) + 11 <= 163840.
+ *
+ *   - On MLU590, the scale needs to meet the following restrictions:
+ *
+ *     - T * (S + 1) + (T + 1) * S + (T + 1) * (S + 1) + 9 * min(S, T) + 11 <= 98304.
+ *
+ * @par API Dependency
+ * - Before calling this function to perform ::mluOpMutualInformationForward, you need to get
+ *   the size of workspace by ::mluOpGetMutualInformationForwardWorkspaceSize.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:7836]:
```suggestion
 *   the size of the workspace by ::mluOpGetMutualInformationForwardWorkspaceSize.
```

File: bangc-ops/mlu_op.h [hunk 5]
```diff
@@ -7698,6 +7699,167 @@ mluOpMutualInformationBackward(mluOpHandle_t handle,
                                const mluOpTensorDescriptor_t py_grad_desc,
                                void *py_grad);
 
+// Group:MutualInformationForward
+/*!
+ * @brief Returns the size of the MLU memory that is used as an extra workspace
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:7704]:
```suggestion
 * @brief Returns the size of the MLU memory as an extra workspace
```
技术英文中是不是句子可以简单一些~

File: bangc-ops/mlu_op.h [hunk 6]
```diff
@@ -7698,6 +7699,167 @@ mluOpMutualInformationBackward(mluOpHandle_t handle,
...
+ * For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] opt_boundary
+ * Pointer to the MLU memory that stores the \b opt_boundary tensor.
+ * @param[in] p_desc
+ * The descriptor of the tensor \b p, which contains dimension, data type, and data layout.
+ * For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] p
+ * Pointer to the MLU memory that stores the tensor \b p.
+ * @param[in] workspace
+ * Pointer to the MLU memory that is used as an extra workspace for the mutual_information_forward operation.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:7790]:
```suggestion
 * Pointer to the MLU memory as an extra workspace for the mutual_information_forward operation.
```

File: bangc-ops/mlu_op.h [hunk 7]
```diff
@@ -7698,6 +7699,167 @@ mluOpMutualInformationBackward(mluOpHandle_t handle,
...
+ * - None.
+ *
+ * @par Data Layout
+ * - None.
+ *
+ * @par Scale Limitations
+ * - None.
+ *
+ * @par API Dependency
+ * - The allocated extra workspace should be passed to ::mluOpMutualInformationForward.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:7741]:
```suggestion
 * - The allocated extra workspace must be passed to ::mluOpMutualInformationForward.
```

File: bangc-ops/mlu_op.h [hunk 8]
```diff
@@ -7698,6 +7699,167 @@ mluOpMutualInformationBackward(mluOpHandle_t handle,
...
+                                              const mluOpTensorDescriptor_t p_desc,
+                                              const mluOpTensorDescriptor_t ans_desc,
+                                              size_t *workspace_size);
+
+// Group:MutualInformationForward
+/*!
+ * @brief Computes mutual information between tensor \b px and tensor \b py.
+ *
+ * @param[in] handle
+ * Handle to an MLUOP context that is used to manage MLU devices and queues in the
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:7767]:
```suggestion
 * Handle to an MLUOP context for MLU devices and queues management in the
```

**Review by AquamarineSu — APPROVED** [review]

**Review by AquamarineSu — APPROVED** [review]

File: bangc-ops/kernels/mutual_information_forward/mutual_information_forward_block.mlu [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+
+__nram__ char nram_buffer[MAX_NRAM_SIZE];
+
+__mlu_func__ void computeLog(float *nram_dst, float *nram_src,
+                             const int32_t deal_num) {
+  int x2d = 0x3f317217;
+  float rlog2e = *(float *)&x2d;
+  __bang_log(nram_dst, nram_src, deal_num);
+  __bang_mul_scalar(nram_dst, nram_src, rlog2e, deal_num);
+}
```
Comment (YunyiGUANG) [inline bangc-ops/kernels/mutual_information_forward/mutual_information_forward_block.mlu:39]: 如果是公共函数，可以像nl中一样，封装到bangc-ops/kernels/utils/common.h里

File: bangc-ops/kernels/mutual_information_forward/mutual_information_forward_block.mlu [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+                               float *max_value, float *mask, float *temp,
+                               int data_num) {
+  __bang_nan_minimum(dst, src1, src2, data_num);
+  __bang_maximum(max_value, src1, src2, data_num);
+
+  // if src1 is nan, then max_value = src1
+  int nan_mask_i = 0x7fffffff;
+  float nan_mask_f = *(float *)&nan_mask_i;
+  __bang_band_scalar(mask, src1, nan_mask_f, data_num);
+  __bang_ge_scalar((int *)mask, (int *)mask, MIN_POS_NAN, data_num);
```
Comment (YunyiGUANG) [inline bangc-ops/kernels/mutual_information_forward/mutual_information_forward_block.mlu:51]: 这条指令应该是操作数含nan时恒返回0，scalar是POS_NAN时这个行为是满足要求的吗？

File: bangc-ops/kernels/mutual_information_forward/mutual_information_forward_block.mlu [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+  __bang_mul_scalar((int *)mask, (int *)mask, -1, data_num);
```
Comment (YunyiGUANG) [inline bangc-ops/kernels/mutual_information_forward/mutual_information_forward_block.mlu:52]: 50-52三行有点没太看明白

File: bangc-ops/kernels/mutual_information_forward/mutual_information_forward_block.mlu [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+  // compute log sum exp
+  __bang_sub(dst, dst, max_value, data_num);
+  __bang_ge_scalar(mask, dst, MIN_LOG_DIFF_FLOAT, data_num);
+  computeExp(dst, dst, nullptr, 0, data_num);
+  __bang_add_scalar(dst, dst, 1.f, data_num);
+  computeLog(dst, dst, data_num);
+  __bang_add(dst, dst, max_value, data_num);
+
+  // if min_value - max_value < MIN_POS_NAN, return the larger one
+  __bang_float2int32_rn((int *)mask, mask, data_num, 0);
```
Comment (YunyiGUANG) [inline bangc-ops/kernels/mutual_information_forward/mutual_information_forward_block.mlu:66]: float2int竞品的rounding mode是tz，虽然这里mask是0/1应该没啥影响，不过最好统一用tz。

File: bangc-ops/kernels/mutual_information_forward/mutual_information_forward_block.mlu [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *************************************************************************/
+#include "mutual_information_forward.h"
+
+#include "kernels/kernel.h"
+#include "kernels/utils/common.h"
+
+#define MIN_LOG_DIFF_FLOAT -15.9423847198486328125f
+#define MIN_POS_NAN  0x7f800001
```
[the remark this answers is not in the document]
Comment (tcd-fang) [reply bangc-ops/kernels/mutual_information_forward/mutual_information_forward_block.mlu:29]: 这里需要使用bit位最小的nan，不确定float.h中定义的宏NAN的bit是啥

File: bangc-ops/kernels/mutual_information_forward/mutual_information_forward_block.mlu [hunk 14]
```diff
@@ -0,0 +1,274 @@
...
+  float rlog2e = *(float *)&x2d;
+  __bang_log(nram_dst, nram_src, deal_num);
+  __bang_mul_scalar(nram_dst, nram_src, rlog2e, deal_num);
+}
+
+__mlu_func__ void logAddVector(float *dst, float *src1, float *src2,
+                               float *max_value, float *mask, float *temp,
+                               int data_num) {
+  __bang_nan_minimum(dst, src1, src2, data_num);
+  __bang_maximum(max_value, src1, src2, data_num);
```
[the remark this answers is not in the document]
Comment (tcd-fang) [reply bangc-ops/kernels/mutual_information_forward/mutual_information_forward_block.mlu:45]: 不需要总是返回nan，即max(a,b)中：
当a=1， b=nan是，返回1；
当a=nan， b=1是，返回nan'；
当a=nan， b=nan是，返回nan；

File: bangc-ops/kernels/mutual_information_forward/mutual_information_forward_block.mlu [hunk 15]
```diff
@@ -0,0 +1,274 @@
...
+  __bang_add_scalar(dst, dst, 1.f, data_num);
+  computeLog(dst, dst, data_num);
+  __bang_add(dst, dst, max_value, data_num);
+
+  // if min_value - max_value < MIN_POS_NAN, return the larger one
+  __bang_float2int32_rn((int *)mask, mask, data_num, 0);
+  __bang_mul_scalar((int *)mask, (int *)mask, -1, data_num);
+  __bang_band((char *)dst, (char *)dst, (char *)mask, data_num * sizeof(float));
+  __bang_add_scalar((int *)mask, (int *)mask, 1, data_num);
+  __bang_mul_scalar((int *)mask, (int *)mask, -1, data_num);
```
Comment (YunyiGUANG) [inline bangc-ops/kernels/mutual_information_forward/mutual_information_forward_block.mlu:70]: 感觉这种naninf刷数操作可以用查找表做，可以参考下面wiki的三.1节看看性能是不是更好。

https://wiki.cambricon.com/pages/viewpage.action?pageId=57471455#id-3)%E6%BF%80%E6%B4%BB(%E8%B6%85%E8%B6%8A%E5%87%BD%E6%95%B0)%E7%B1%BB%E7%AE%97%E5%AD%90%E7%B2%BE%E5%BA%A6%E9%97%AE%E9%A2%98%E6%80%BB%E7%BB%93-1.%E7%89%B9%E6%AE%8A%E4%BD%8D%E7%BD%AE%E5%88%B70/%E5%85%B6%E4%BB%96%E5%9B%BA%E5%AE%9A%E5%80%BC%E7%9A%84%E5%A4%84%E7%90%86

File: bangc-ops/kernels/mutual_information_forward/mutual_information_forward_block.mlu (hunk 11 above)
Comment (tcd-fang) [reply bangc-ops/kernels/mutual_information_forward/mutual_information_forward_block.mlu:52]: 45行max_value表示src1和src2中的最大值(不敢值nan)，48-55行的意思是如果src1某些位置为nan，那么max_value对应位置也要刷成nan。
45行不用__bang_nan_maximum的原因是，当src1=1，src2=nan，则max_value=nan。但竞品在此情况下，max_value=1

File: bangc-ops/kernels/mutual_information_forward/mutual_information_forward.cpp [hunk 16]
```diff
@@ -0,0 +1,432 @@
...
+               << " Current operator supports tensor num smaller than 2^31.";
+    return MLUOP_STATUS_NOT_SUPPORTED;
+  }
+
+  const int S = px_desc->dims[1];
+  const int T = py_desc->dims[2];
+
+  // check scale limit for compute p
+  int currnet_size = T * (S + 1) + (T + 1) * S + (T + 1) * (S + 1) +
+                     9 * std::min(S, T) + 11;
```
Comment (YunyiGUANG) [inline bangc-ops/kernels/mutual_information_forward/mutual_information_forward.cpp:222]: 9和11最好加下注释

File: bangc-ops/kernels/mutual_information_forward/mutual_information_forward.cpp [hunk 17]
```diff
@@ -0,0 +1,432 @@
...
+  PARAM_CHECK(API_NAME, px_desc != nullptr);
+  PARAM_CHECK(API_NAME, py_desc != nullptr);
+  PARAM_CHECK(API_NAME, p_desc != nullptr);
+  PARAM_CHECK(API_NAME, ans_desc != nullptr);
+
+  // since the layout of all tensor is ARRAY, so skip check tensor layout
+
+  // 2. check mlu platform
+  if (handle->arch < 372) {
+    LOG(ERROR) << API_NAME << " Only mlu300 and above devices are supported."
```
Comment (YunyiGUANG) [inline bangc-ops/kernels/mutual_information_forward/mutual_information_forward.cpp:289]: 322也是300系列的卡，如果不支持322，这里报错信息最好对齐下。

File: bangc-ops/kernels/mutual_information_forward/mutual_information_forward_block.mlu (hunk 9 above)
Comment (tcd-fang) [reply bangc-ops/kernels/mutual_information_forward/mutual_information_forward_block.mlu:39]: 如果写在common.h里，还需要兼容一下200系列的激活表。暂时先不放在common.h里

File: bangc-ops/mlu_op.h [hunk 18]
```diff
@@ -7698,6 +7699,167 @@ mluOpMutualInformationBackward(mluOpHandle_t handle,
...
+     If \b opt_boundary is NULL, it will be treated as [0, 0, S, T].
+ * - The shape of \b p is 3D([B, S + 1, T + 1]).
+ * - The shape of \b ans is 1D([B]).
+ * - The size of each tensor must be in the range of [0, 2^31].
+ *
+ *   - On MLU370, the scale needs to meet the following restrictions:
+ *
+ *     - T * (S + 1) + (T + 1) * S + (T + 1) * (S + 1) + 9 * min(S, T) + 11 <= 163840.
+ *
+ *   - On MLU590, the scale needs to meet the following restrictions:
```
[the remark this answers is not in the document]
Comment (tcd-fang) [reply bangc-ops/mlu_op.h:7830]: 上次合入mutual_information_backward是，沟通过。可以写590，不暴露相关参数即可

File: bangc-ops/kernels/mutual_information_forward/mutual_information_forward.cpp (hunk 16 above)
Comment (tcd-fang) [reply bangc-ops/kernels/mutual_information_forward/mutual_information_forward.cpp:224]: 已添加

**Review by YunyiGUANG — APPROVED** [review]

File: docs/bangc-docs/api_guide/update.rst [hunk 19]
```diff
@@ -3,6 +3,17 @@ Update History
...
+* V0.7.0
+
+  **Date:** June 2, 2023
+
+  **Changes:**
+
+  - Added the following new operations:
+
+    - mutual_information_backward
+    - mutual_information_forward
```
Comment (duzekunKTH) [inline docs/bangc-docs/api_guide/update.rst:15]: 这个文件不需要开发者修改，请删除

File: docs/bangc-docs/user_guide/2_update_history/index.rst [hunk 20]
```diff
@@ -1,5 +1,16 @@
...
+* **V0.7.0**
+
+  **更新时间**：2023年6月2日
+
+  **更新内容**：
+
+  - 新增以下算子章节内容：
+
+     + :ref:`mutual_information_backward`
+     + :ref:`mutual_information_forward`
```
Comment (duzekunKTH) [inline docs/bangc-docs/user_guide/2_update_history/index.rst:12]: 此文件不需要开发者修改，开发者只需要更新 [docs/bangc-docs/user_guide/9_operators/index.rst](https://github.com/Cambricon/mlu-ops/blob/master/docs/bangc-docs/user_guide/9_operators/index.rst)

File: docs/bangc-docs/api_guide/update.rst (hunk 19 above)
Comment (tcd-fang) [reply docs/bangc-docs/api_guide/update.rst]: 已删除

File: docs/bangc-docs/user_guide/2_update_history/index.rst (hunk 20 above)
Comment (tcd-fang) [reply docs/bangc-docs/user_guide/2_update_history/index.rst]: 已删除

**Review by alex-xuwenming — APPROVED** [review]

**Pull request merged**